arch/xtensa/soc/esp32/esp32-mp.c

```diff
@@ -0,0 +1,190 @@
+	 * WINDOWSTART and will never be asked to spill anything.
+	 * Those 16 bytes would otherwise be wasted on the stack, so
+	 * adjust */
+	"addi a1, a2, 16"	"\n\t"
+
+        /* Clear WINDOWSTART so called functions never try to spill
+         * our callers' registers into the now-garbage stack pointers
+         * they contain.  No need to set the bit corresponding to
+         * WINDOWBASE, our C callee will do that when it does an
+         * ENTRY. */
+        "movi a0, 0"		"\n\t"
+        "wsr.WINDOWSTART a0"	"\n\t"
+
+        /* Clear CALLINC field of PS (you would think it would, but
+         * our ENTRY doesn't actually do that) so the callee's ENTRY
+         * doesn't shift the registers */
+        "rsr.PS a0"		"\n\t"
+        "movi a2, 0xfffcffff"	"\n\t"
+        "and a0, a0, a2"	"\n\t"
+        "wsr.PS a0"		"\n\t"
+
+        "rsync"			"\n\t"
+        "movi a0, 0"		"\n\t"
+
+	"jx a3"			"\n\t");
+
+/* Carefully constructed to use no stack beyond compiler-generated ABI
+ * instructions.  WE DO NOT KNOW WHERE THE STACK FOR THIS FUNCTION IS.
+ * The ROM library just picks a spot on its own with no input from our
+ * app linkage and tells us nothing about it until we're already
+ * running.
+ */
+static void appcpu_entry1(void)
+{
+	_appcpu_stack_switch(appcpu_top, appcpu_entry2);
+}
+
+/* The calls and sequencing here were extracted from the ESP-32
+ * FreeRTOS integration with just a tiny bit of cleanup.  None of the
+ * calls or registers shown are documented, so treat this code with
+ * extreme caution.
+ */
+static void appcpu_start(void)
+{
+	/* These two calls are wrapped in a "stall_other_cpu" API in
+	 * esp-idf.  But in this context the appcpu is stalled by
+	 * definition, so we can skip that complexity and just call
+	 * the ROM directly.
+	 */
+	esp32rom_Cache_Flush(1);
+	esp32rom_Cache_Read_Enable(1);
+
+	RTC_CNTL_SW_CPU_STALL &= ~RTC_CNTL_SW_STALL_APPCPU_C1;
+	RTC_CNTL_OPTIONS0     &= ~RTC_CNTL_SW_STALL_APPCPU_C0;
+	DPORT_APPCPU_CTRL_B   |= DPORT_APPCPU_CLKGATE_EN;
+	DPORT_APPCPU_CTRL_C   &= ~DPORT_APPCPU_RUNSTALL;
+
+	/* Pulse the RESETTING bit */
+	DPORT_APPCPU_CTRL_A |= DPORT_APPCPU_RESETTING;
+	DPORT_APPCPU_CTRL_A &= ~DPORT_APPCPU_RESETTING;
+
+	/* Seems weird that you set the boot address AFTER starting
+	 * the CPU, but this is how they do it...
+	 */
+	esp32rom_ets_set_appcpu_boot_addr((uint32_t)appcpu_entry1);
+}
+
+void _arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+                     void (*fn)(int, void *), void *arg)
+{
+	volatile struct cpustart_rec sr;
+	int vb;
+	volatile int alive_flag;
+
+	__ASSERT(cpu_num == 1, "ESP-32 supports only two CPUs");
+
+	__asm__ volatile("rsr.VECBASE %0\n\t" : "=r"(vb));
+
+	alive_flag = 0;
+
+	sr.cpu = cpu_num,
+	sr.fn = fn,
+	sr.stack_top = K_THREAD_STACK_BUFFER(stack) + sz,
+	sr.arg = arg,
+	sr.vecbase = vb,
+	sr.alive = &alive_flag,
+
+	appcpu_top = K_THREAD_STACK_BUFFER(stack) + sz;
+
+	start_rec = &sr;
```

Comment:
Will sr be alive when _arch_start_cpu() ends?

Reply:
Yes.  Fundamentally that struct is just a bunch of arguments passed to the handler function on the other CPU.  It exists on the stack here so we don't have to allocate extra static memory for it.  There's an explicit loop in the owning function to spin on the other CPU being finished.